File: python/tsp_sequential.py
#!/usr/bin/env python3
"""tsp_sequential.py - Task 5: Chạy GA tuần tự trên 1 process cho 1 bài TSP.

Demo: python3 tsp_sequential.py ../data/cities_30.txt --gens 500 --pop 200
In ra độ dài tour tốt nhất và lưu lộ trình ra file (tùy chọn).
"""
import argparse
import time
import numpy as np
import ga_core as ga


def main():
    ap = argparse.ArgumentParser(description="GA tuan tu cho TSP")
    ap.add_argument("cities", help="file toa do thanh pho")
    ap.add_argument("--gens", type=int, default=500, help="so the he")
    ap.add_argument("--pop", type=int, default=200, help="kich thuoc quan the")
    ap.add_argument("--seed", type=int, default=42)
    ap.add_argument("--out", default=None, help="file luu tour tot nhat")
    args = ap.parse_args()

    coords = ga.read_cities(args.cities)
    D = ga.distance_matrix(coords)
    rng = np.random.default_rng(args.seed)

    t0 = time.time()
    tour, length, history = ga.evolve(D, args.pop, args.gens, rng)
    dt = time.time() - t0

    print(f"So thanh pho   : {len(coords)}")
    print(f"The he         : {args.gens}, quan the: {args.pop}")
    print(f"Do dai tot nhat: {length:.2f}")
    print(f"Thoi gian       : {dt:.2f}s")
    print(f"Lo trinh        : {tour.tolist()}")

    if args.out:
        np.savetxt(args.out, tour, fmt="%d")
        print(f"Da luu tour vao {args.out}")


if __name__ == "__main__":
    main()

File: cpp/tsp_sequential.cpp
// tsp_sequential.cpp - Task 5: Run sequential GA (C++) for one TSP instance.
// Compile: g++ -O2 -o tsp_seq tsp_sequential.cpp
// Run:     ./tsp_seq ../data/cities_30.txt 300 150 42
#include "ga_core.hpp"
#include <iostream>
#include <chrono>

int main(int argc, char** argv) {
    if (argc < 2) { std::cerr << "Usage: " << argv[0]
        << " <file_cities> [gens] [pop] [seed]\n"; return 1; }
    std::string path = argv[1];
    int gens = argc > 2 ? std::stoi(argv[2]) : 500;
    int pop_size = argc > 3 ? std::stoi(argv[3]) : 200;
    unsigned seed = argc > 4 ? std::stoul(argv[4]) : 42;

    auto coords = read_cities(path);
    int n = (int)coords.size();
    auto D = distance_matrix(coords);
    std::mt19937 rng(seed);

    std::vector<Tour> pop(pop_size);
    std::vector<double> len(pop_size);
    for (int i = 0; i < pop_size; i++) {
        pop[i] = random_tour(n, rng);
        len[i] = tour_length(pop[i], D, n);
    }

    auto t0 = std::chrono::high_resolution_clock::now();
    for (int g = 0; g < gens; g++)
        evolve_one_gen(pop, len, D, n, 1, 5, 0.3, rng);
    auto t1 = std::chrono::high_resolution_clock::now();

    int best = (int)(std::min_element(len.begin(), len.end()) - len.begin());
    double secs = std::chrono::duration<double>(t1 - t0).count();
    std::cout << "Cities         : " << n << "\n"
              << "Generations    : " << gens << ", population: " << pop_size << "\n"
              << "Best length    : " << len[best] << "\n"
              << "Time            : " << secs << "s\n";
    std::cout << "Route           :";
    for (int c : pop[best]) std::cout << " " << c;
    std::cout << "\n";
    return 0;
}
